semgrep/semgrep/core_runner.py

```diff
@@ -225,111 +291,93 @@ def _run_rule(
-            # performed purely in Python code then compared against
-            # semgrep-core's results for other patterns.
-            patterns_regex, patterns = partition(
-                lambda p: p.expression.operator == OPERATORS.REGEX,
-                all_patterns_for_language,
-            )
-            if patterns_regex:
-                patterns_json = [pattern.to_json() for pattern in patterns_regex]
-
-                try:
-                    patterns_re = [
-                        (pattern["id"], re.compile(pattern["pattern"]))
-                        for pattern in patterns_json
-                    ]
-                except re.error as err:
-                    raise SemgrepError(f"invalid regular expression specified: {err}")
-
-                re_fn = functools.partial(get_re_matches, patterns_re)
-                with multiprocessing.Pool(self._jobs) as pool:
-                    matches = pool.map(re_fn, targets)
-
-                outputs.extend(
-                    single_match
-                    for file_matches in matches
-                    for single_match in file_matches
+                pattern_json = rule._raw.copy()
+                del pattern_json["mode"]
+                pattern = Pattern(
+                    0, rule.expression, rule.severity, language, rule._yaml.span
                 )
 
-            patterns_json = [p.to_json() for p in patterns]
-            with tempfile.NamedTemporaryFile(
-                "w"
-            ) as pattern_file, tempfile.NamedTemporaryFile(
-                "w"
-            ) as target_file, tempfile.NamedTemporaryFile(
-                "w"
-            ) as equiv_file:
-                yaml = YAML()
-                yaml.dump({"rules": patterns_json}, pattern_file)
-                pattern_file.flush()
-                target_file.write("\n".join(str(t) for t in targets))
-                target_file.flush()
-
-                cmd = [SEMGREP_PATH] + [
-                    "-lang",
+                output_json = self._run_core_command(
+                    [pattern_json],
+                    [pattern],
+                    targets,
                     language,
-                    "-rules_file",
-                    pattern_file.name,
-                    "-j",
-                    str(self._jobs),
-                    "-target_file",
-                    target_file.name,
-                    "-use_parsing_cache",
+                    rule,
+                    "-tainting_rules_file",
                     cache_dir,
-                ]
-
-                if equivalences:
-                    self._write_equivalences_file(equiv_file, equivalences)
-                    cmd += ["-equivalences", equiv_file.name]
+                )
 
-                core_run = sub_run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        else:
+            for language, all_patterns_for_language in self._group_patterns_by_language(
```

Comment:
This function is probably complex enough that we want to subfunction it.

I can show you a fun PyCharm-fu to automate this.